libzeth/circuits/circuit_utils.tcc

```diff
@@ -46,22 +46,21 @@ libsnark::linear_combination<FieldT> packed_addition(
     // Thus here, we make sure our binary string is interpreted correctly.
     return libsnark::pb_packing_sum<FieldT>(
         libsnark::pb_variable_array<FieldT>(inputs.rbegin(), inputs.rend()));
-};
+}
 
-// Takes a vector of boolean values, and convert this vector of boolean values
-// into a vector of FieldT::zero() and FieldT:one()
+// Allocate an array of variables on a given protoboard, and set the values to
+// zero or one based on a vector of bits.
 template<typename FieldT>
 libsnark::pb_variable_array<FieldT> variable_array_from_bit_vector(
```

Comment:
Actually, since the `fill_with_bits` method of `pb_variable_array` is used here, can't we simply remove the: `bits<numBits>::fill_variable_array` function? Looks like it is redundant since every call of `bits<numBits>::fill_variable_array` will be of the form: `<bit_vector>.fill_variable_array(pb, <pb_variable_array>);` and can thus be expressed by switching the pb_var and the bit vector as `<pb_variable_array>.fill_with_bits(pb, <bit_vector>);` 

Do you remember why the `bits<numBits>::fill_variable_array` function was needed (@dtebbs)?

Note: If still needed it would probably be more appropriate to rename `fill_variable_array` to `fill_pb_variable_array` to stick to type names, but the changes related to this function would need to be done in another PR anyway since it goes beyond the scope of this specific PR (`bits.fill_variable_array` is used in many gadgets)

Reply:
For now, `bits<numBits>` is not compatible with `pb_variable_array<FieldT>::fill_with_bits`, so we would have to overload or modify this function in order to replace `bits<numBits>::fill_variable_array`.
Or, it could make sense to create an equivalent function to `variable_array_from_bit_vector` in this repo.

As you say, it would touch a lot of code, so maybe one for another PR.

Reply:
>> Do you remember why the bits<numBits>::fill_variable_array function was needed (@dtebbs)?
> For now, bits<numBits> is not compatible with pb_variable_array<FieldT>::fill_with_bits

Ah yes the types won't match..
From what I can see, `fill_with_bits`: https://github.com/clearmatics/libsnark/blob/master/libsnark/gadgetlib1/pb_variable.tcc#L49-L57 and `fill_variable_array`: https://github.com/clearmatics/zeth/blob/develop/libzeth/core/bits.tcc#L107-L119 are running the same code. Only diff is indeed on the type side of things, one is manipulating a vector, the other is manipulating an array. So `bits<numBits>::fill_variable_array` can be expressed as the composition of `array_to_vector` and `pb_variable_array<FieldT>::fill_with_bits` (i.e. its definition can be expressed as a wrapper around this function composition - introducing 2 additional function calls and context switching. This code is far from being the bottleneck though, so no major deal here). Anyway, not a big one, but I think we should rename the functions to strictly stick to the `<typeX>_to_<typeY>` syntax, and maybe consider some minor refactoring on the `bits` class code and/or `pb_variable` code. One for future PRs.

Reply:
> So bits<numBits>::fill_variable_array can be expressed as the composition of array_to_vector and pb_variable_array<FieldT>::fill_with_bits (i.e. its definition can be expressed as a wrapper around this function composition - introducing 2 additional function calls and context switching. This code is far from being the bottleneck though, so no major deal here)

For the future but just to note, composing those functions would mean allocating and copying these structures, no matter how larger they are, just to deal with type differences.  A generic `fill_with_bits` function that can accept different containers would allow us to combine the code and avoid unnecessary allocating and copying.

Reply:
> A generic fill_with_bits function that can accept different containers would allow us to combine the code and avoid unnecessary allocating and copying.

Oh yes definitely, this would be much better